packages/client-python/src/rocketride/pipeline.py

```
@@ -0,0 +1,538 @@
+
+        Returns:
+            self, for fluent chaining.
+
+        Raises:
+            ValueError: If ``node_id`` already exists or a second source is added.
+        """
+        if node_id in self._nodes:
+            raise ValueError(f"Node '{node_id}' already exists in the pipeline")
+
+        if source and self._source_id is not None:
+            raise ValueError(f"Pipeline already has a source node '{self._source_id}'. Cannot add '{node_id}' as a second source.")
```

Ruff (0.15.7)

[warning] Avoid specifying long messages outside the exception class

(TRY003)